### ISSUE TITLE
IndexError: tuple index out of range

### ISSUE THREAD

**Issue by alexnikitchuk** [sqlfluff/sqlfluff#3235]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When I try to run linter against the query I get `IndexError: tuple index out of range` error.

### Expected Behaviour

no errors

### Observed Behaviour

```
WARNING    Unable to lint models/marts/dispatching/game_driver.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!                                            
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/linter/runner.py", line 103, in run
    yield partial()
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/linter/linter.py", line 743, in lint_rendered
    parsed = cls.parse_rendered(rendered)
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/linter/linter.py", line 374, in parse_rendered
    parsed, pvs = cls._parse_tokens(
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/linter/linter.py", line 207, in _parse_tokens
    parsed: Optional[BaseSegment] = parser.parse(
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/parser.py", line 36, in parse
    parsed = root_segment.parse(parse_context=ctx)
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/segments/base.py", line 1086, in parse
    self.segments = self.expand(
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/segments/base.py", line 438, in expand
    res = stmt.parse(parse_context=parse_context)
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/segments/base.py", line 1086, in parse
    self.segments = self.expand(
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/segments/base.py", line 438, in expand
    res = stmt.parse(parse_context=parse_context)
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/segments/base.py", line 1032, in parse
    check_still_complete(segments, m.matched_segments, m.unmatched_segments)
  File "/usr/local/lib/python3.9/site-packages/sqlfluff/core/parser/helpers.py", line 34, in check_still_complete
    segment=unmatched_segments[0],
IndexError: tuple index out of range
 
All Finished 📜 🎉!
```

### How to reproduce

run `sqlfluff lint` against query 
```
select
    value:point:from:latitude::NUMBER(10, 6) as lat,
    value:point:from:longitude::NUMBER(10, 6) as lng
from {{ ref('table') }}
```

### Dialect

snowflake

### Version

sqlfluff, version 0.13.0

### Configuration

.sqlfluff
```
[sqlfluff]
templater = jinja
dialect = snowflake
exclude_rules = L031

[sqlfluff:indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false

[sqlfluff:templater]
unwrap_wrapped_queries = true

[sqlfluff:templater:jinja]
apply_dbt_builtins = true

[sqlfluff:rules]
max_line_length = 180

[sqlfluff:rules:L029]
# Keywords should not be used as identifiers.
# Comma separated list of words to ignore for this rule
ignore_words = comment,type,version,timestamp,name,model

[sqlfluff:rules:L010]
# Inconsistent capitalisation of keywords
capitalisation_policy = lower

[sqlfluff:rules:L014]
# Inconsistent capitalisation of unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:L030]
# Inconsistent capitalisation of function names
extended_capitalisation_policy = upper

[sqlfluff:rules:L063]
# Inconsistent capitalisation of datatypes
extended_capitalisation_policy = upper

[sqlfluff:rules:L031]
# Avoid table aliases in from clauses and join conditions.
force_enable = False
```

.sqlfluffignore
```
target/
# dbt <1.0.0
dbt_modules/
# dbt >=1.0.0
dbt_packages/
macros/
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
This is unparsable because of the `from` in the column names. Is that valid SQL since `FROM` is a reserved word in Snowflake.

Still shouldn't throw that horrible error - that's easy to fix.

**Comment by alexnikitchuk** [comment]
Snowflake doesn't complain about the query.

**Comment by tunetheweb** [comment]
Weirdly it's fine if there's only one of these select elements using `from`. Or if either (or both) use `from1` instead. Odd. Still digging...

**Comment by tunetheweb** [comment]
There's something up with `enforce_whitespace_preceding_terminator` as when I replace `SelectClauseSegment` `match_grammar with this:

```py
    match_grammar = StartsWith(
        "SELECT",
        terminator=OneOf(
            Sequence("FROM", Ref("FromExpressionSegment")),
            "WHERE",
            Sequence("ORDER", "BY"),
            "LIMIT",
            "OVERLAPS",
            Ref("SetOperatorSegment"),
            "FETCH",
            "OFFSET",
        ),
        enforce_whitespace_preceding_terminator=True,
    )
```

It works. But if the first one is just `"FROM"` then it doesn't work, even though there is no whitespace preceding the terminator.

**Comment by tunetheweb** [comment]
OK found the bug and full fix is in #3236 . Will be included in next release.
